keyup / keydown binding

bovine3dom opened the pull request:
Need to think about how to add this to the grammar. Toying with `<C-:KeyUp:j>`. I think I'll add `:KeyDown:` as a no-op/default. 

Closes #1626.

bovine3dom committed:
WIP keyup binding

bovine3dom commented on src/grammars/bracketexpr.ne:
```diff
@@ -11,6 +12,7 @@ Modifier -> [acmsACMS]:? [acmsACMS]:? [acmsACMS]:? [acmsACMS]:? "-" {%
             ["M", "metaKey"],
             ["S", "shiftKey"],
+            /** Will need to add KeyUp here */
         ])
```
I'm very tempted to just use U and D since they aren't used for anything else.

bovine3dom commented:
One issue we'll need to work around is that each KeyUp event will break any multi-letter binds, e.g. g; instead gets g then keyup-g, so g; is never matched

bovine3dom committed:
Add working keyup parsing

NB: this breaks special keys e.g. `<Enter>`

bovine3dom commented:
This currently breaks `<Enter>` etc so the command line can't be used. Can be tested with the address bar - e.g. `tri bind <U-j> scrollline -5`.

bovine3dom committed:
Add explicit non-repeat keydown bindings

bovine3dom commented:
Probably worth looking at https://github.com/tridactyl/tridactyl/pull/1814/files to see more places we need to add keyup/down.

bovine3dom committed:
Lazy fix for hint/gobble mode

bovine3dom commented:
I'd like some input on how to proceed with respect to "each KeyUp event will break any multi-letter binds".

I see three options:

- keyup binds are only allowed at the end of a key sequence and are otherwise ignored
    - pros: simple to implement
    - cons: needs enforcement at bind time or users will get confused, prevents complex chorded binds (e.g. if some lunatic wants to implement a stenographic keyboard in Tridactyl)
- convert all un-annotated binds to require a keyup (at parse time), e.g. `gg` -> `<D-g><U-g><D-g><U-g>`. Introduce another annotation, `<R-` to denote a key that may (or may not) be repeat firing (from holding the key down) so that holding down "j" to scroll still works. 
    - pros: allows for complex binds
    - cons: complex to implement. Would break two kinds of behaviour: pressing and holding g then pressing t wouldn't change tab; holding g down until the repeat fired wouldn't trigger gg.
- as above but add another annotation, `<?` which denotes an optional key, so `gg` becomes `<R-g><?U-g><R-g><?U-g>`.
   - pros: doesn't break any existing behaviour, I think
   - cons: hard to implement, but not many others

I think I'll probably go with option 3.

bovine3dom committed:
Add annotation for optional key

bovine3dom committed:
Remove unused declaration

bovine3dom commented:
Before merging:

- [ ] fix binds like `<Enter>` (but only in the command line? seems to work fine elsewhere)
- [ ] check required ordering for `unbind` and whether `bind` reorders them correctly
- [ ] check keytranslatemap behaviour
- [ ] convert binds with no `UDR` modifiers to `<[modifiers]D-x><[modifiers]?U-x>`, https://github.com/tridactyl/tridactyl/blob/5823726ba1de886c715179c1d4ebc9f3cbc36d4c/src/lib/keyseq.ts#L380 seems like a convenient place.

bovine3dom commented:
Just want to document some thoughts on smooth scrolling with this after a brief conversation with hpfr on Riot. We could:

- bind keydown to start scrolling. Bind keyup to stop scrolling. (Videogame style), or
- bind keydown to smoothly scroll a line. bind repeat-keydown to initiate + continue continuous smooth scrolling. Not quite sure how this would work in practice.

glacambre reviewed (commented):
I'm a bit afraid of this change but I don't have any technical argument against it. Turning keyevents into their `<D-*><U-*>` representation sounds sound to me. I'm slightly concerned that it might make Tridactyl slower in a perceivable way but it's probably fine in practice.

bovine3dom commented on src/lib/keyseq.ts:
```diff
@@ -52,12 +55,23 @@ export class MinimalKey {
             // Don't check shiftKey for normal keys.
             if (attr === "shiftKey" && this.key.length === 1) continue
-            if (this[attr] !== keyevent[attr]) return false
+            if (attr === "keydown") {
```
To avoid editing the real event we can add keyup handling here in a similar fashion

bovine3dom committed:
Add R annotation for explicit may-be-repeat

NB: this is actually a placebo, only D binds cannot be repeat firings.
The purpose of this is to allow us to convert bare keys - e.g. g - to
<R-g><?U-g> at parse-time and to allow people to disallow keyup by
explicitly binding to <R-g>.

bovine3dom committed:
Fix daft bug with optional keys

bovine3dom commented:
Since doing it at parse-time is too slow, I think the best way of proceeding is this:

- update config.ts so that, e.g. `gt` becomes `g<?U-g>t`,
- a config updater that updates custom binds in the same way, e.g. `<C-b>t` ->  `<C-b><?UC-b>t`
- change `bind` to bind any key without an explicit `R, U or D` to `<[mods]R-[key]><[mods]?U-[key]>`, while ensuring we leave the final key alone
- ditto for `unbind`.

The obvious downside of this is that `viewconfig nmaps` and `bind gt` (with no bind provided) becomes harder to understand. My favourite so far is `"<<?U-<><": tabmove -1`.

glacambre commented:
This is surprising to me, I wouldn't have expected the parsing stage to 
be the bottleneck. Do you know why it's slow? Is it because we have 
greedy regexes somewhere?

bovine3dom commented:
We convert the whole bit of config, e.g. `nmaps` to `MinimalKeys` for every single keypress. I don't think it's that an extra `new MinimalKey` is particularly slow - it's more that we're doing it _a lot_.

We could probably improve the responsiveness of Tridactyl a lot by doing this just once at startup and if the bit of configuration changed. We'd have to think a bit to support user-defined modes. Any config ending with "maps" perhaps?

bovine3dom committed:
Add optional keyup binds to all simple non-final keys in bound keysequences

E.g. `gg` -> `g<?U-g>`

bovine3dom committed:
Add implicit keyup bind to more complex keys

bovine3dom committed:
Fix special keys not working in command line

bovine3dom committed:
Document keyup, keydown and optional binds

bovine3dom commented:
I think this is ready for review. I'll leave it a few days and have a go myself, but others are welcome.

I should update the unit tests. I'm not very keen on the current ones. I might bin them and switch to testing the parsers instead. (e.g. `isTrusted: false` keys get rejected, see #1833).

bovine3dom committed:
Update some keyseq tests

bovine3dom committed:
Minor refactor

bovine3dom committed:
Fix counts in keyseq

The tests found a real bug. Are they supposed to do that?

bovine3dom committed:
Update more keyseq tests for keyup binds

bovine3dom committed:
Comment out failing keyseq test

bovine3dom commented:
Found a bug: `<Space>` can't be bound to (it doesn't work e.g. to insert a completion in the command line).

bovine3dom committed:
Fix typo

bovine3dom commented:
Failing test:

```
const k = keyseq.mapstrToKeyseq("<Space>")
const km = keyseq.mapstrMapToKeyMap(new Map([["<Space>", "blah"]]))
keyseq.parse(k, km).isMatch == true // but it's false
```

bovine3dom commented:
`modeindicatorshowkeys` shows keydown and keyup prefixes with this PR, but with the keyup/keydown information ignored.

I _think_ the best thing to do is to ignore the keyup keys in the mode indicator - the alternative of `<D-g><U-g>` etc. seems like it would be too ugly for something that 99% of people are never going to care about.

bovine3dom committed:
Merge branch 'master' into pushykeyseq

bovine3dom committed:
Fix keyup events triggering, update tests

bovine3dom committed:
Fix <Space> binds

```
Number(' ') === 0
```

js needs to be nuked from orbit

bovine3dom committed:
Fix hint mode

bovine3dom committed:
Cheer up linter

bovine3dom committed:
Fix regression with repeat keys not firing

bovine3dom committed:
Fix explicit <D-j> matching repeat keys

GHolk committed:
Fix bracket-expr + single char binds

bovine3dom committed:
Fix tests

bovine3dom committed:
Fix <C-,>d bind

bovine3dom committed:
Cheer up linter

bovine3dom committed:
Fix binds greedily matching when there are modifiers

bovine3dom committed:
Warn when keydown binds are shadowed

Main fix is changing from a string match to a check of
what the mapstrs actually parse to

bovine3dom committed:
Fix default nmode binds

bovine3dom committed:
Fix modeindicatorshowkeys

U/D annotations are only shown if matching
bind has them explicitly

bovine3dom committed:
Fix broken examples in docstring

bovine3dom committed:
Merge branch 'master' into pushykeyseq

bovine3dom committed:
Ignore mergiraf orig files

bovine3dom committed:
Do some code golf

bovine3dom committed:
Merge branch 'master' into pushykeyseq

bovine3dom committed:
Document gholk's ignore layer

bovine3dom committed:
Only buffer keydowns for commandline

bovine3dom committed:
Extract queue management from cancelling

This stops us from having to cancel keyups twice

bovine3dom committed:
Make sure keypress events aren't cancelled twice

bovine3dom committed:
Stop mutating vanilla keyevent

bovine3dom committed:
Improve docs for legacy modes

bovine3dom committed:
Add unit test to ensure untrusted keys are rejected
